### ISSUE TITLE
Discussion about our issue template's default labels

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/team-compass#334]
When a person creates a bug report, we have [declared organization wide issue templates](https://github.com/jupyterhub/.github) that adds a "bug" or "enhancement" label.

I'd like for us to switch the use of a "bug" label to bug/possible and bug/confirmed or something similar, [like the Traefik repo does](https://github.com/containous/traefik/issues) for example.

Similarly, for the enhancement issue, it would be good to switch the use of a "enhancement" by default to a label of enhancement/idea which can transition to enhancement/specified-and-agreed or similar.

I think this would help direct attention for both browsing users and active contributors, and for me personally, I find it valuable to understand if another maintainer has concluded if something is an actual bug or if an enhancement idea makes enough sense to develop.

What do you think?

**Comment by manics** [comment]
For bugs how about `needs triage` since it's clear, especially to new contributors, what action is required?

**Comment by GeorgianaElena** [comment]
The reason why the labels were added was to make it easier to spot the support issues (ref: https://github.com/jupyterhub/.github/pull/6).
But it looks like when something is not working people might think is because of a bug and use the bug template, which makes it harder to distinguish between bug reports from support issues.

So, adding one extra layer of labels to help triage them, sounds like a very good idea.

**Comment by minrk** [comment]
Great idea! I think `needs: triage` is a good starting condition for bug reports, and `idea` or `needs: discussion` on feature requests makes sense.

I agree that there's a challenge for filtering Issues as a task list based on what their need is and what your capacity is. I have found issue classification to be super hard. I tried to have thorough tagging for a long time on jupyter/notebook but never managed to keep it up to date, and eventually gave up on it. Without rigorous tagging, filtering based on tags is hard because you can't trust the tags to be showing you all the issues in the category you are interested in. It can still be helpful if it's strictly additive information, though, i.e. issues with this tag are *definitely* in this category, and issues without it *may* be in the same category.

I have found that the distinction between "bug" and "enhancement" has rarely if ever been a useful one. It is rarely relevant whether a person's problem is caused by a missing feature or a bug—they have a need that is not addressed. New features may be simple additions and really important, while certain bugs may be really complicated and relatively unimportant or impossible to fix without backward-incompatible changes. For this reason, I'd eliminate the bug/enhancement distinction and try to label with what is needed. We don't currently have a label for "discussion resolved, needs someone's time to implement the solution", and it sounds like that's one of the classifications @consideRatio is after, right? Perhaps `needs: implementation`? Would someone like to try an experiment of finding current Issues that are in this state?

Maybe it would be good to collect some examples of "I am contributor X, I can do Y, have Z amount of time, and want to find something to work on" and how we might be able to classify issues to help with some of those? It's only useful to have tags for things where the distinction from not having that tag results in people making different decisions about that Issue, whether that's skipping over it for now, or diving in.

Some possibly useful distinctions that can help folks determine where they should spend their time:

- new issue, needs initial classification (`needs: triage` proposed above, `idea` is similar but represents a different kind of discussion and time-scale than a bug report)
- is a support request / discussion that (probably) won't be tied to any changes in code/docs (we have `question` for this, mostly, and are trying to push these to discourse for the most part)
- needs more info from the OP, such as bug confirmation, system info, requirements, etc. (`needs: info` used in some repos)
- needs more discussion to determine what should be done before implementing it (we have `needs: discussion`)
- needs input, expertise, or time not likely to come from the maintainers (`help wanted` tag is used for this sometimes)
- someone is already actively working on this
- needs implementation, and
  - is a deep, sensitive refactor that touches lots of things (not many contributors are comfortable with these)
  - has low conflict with other tasks (this is a good criterion for an issue being new-contributor friendly)
  - is relatively simple and localized (more tractable, maybe you don't have much time to work on something)
  - requires a backward-incompatible change (sensitive to release cycle)
  - has to wait for X other task (nothing to do now, but may prioritize X)
  - involves the database, tornado, jinja, javascript, CI/CD or other component (different folks have different expertise and comfort)

I'm not proposing that we have tags for all of those things, just trying to come up with different sources of information that can help people decide whether an issue is one for them to pick up or move past.

The JupyterLab team has also done some work on [pretty detailed issue classification](https://github.com/jupyter/governance/pull/29), which they have found helpful given their high volume of issues, but I think that approach is too much work to keep up to date.

The hardest part of issue classification for me is balancing useful classifications so people can find things with the time required to keep up with those classifications.

**Comment by manics** [comment]
If we had `needs info` would it be useful to have a bot that replaces the label with something to indicate it's ready for reassessment?

It would also be good if there's a way to mark an issue as re-triaged by us. For example when I've gone through a pile of old issues on some repos I might decide they should be kept open, but it'd be a waste of time for someone else to re-assess the same issues, when there are 100s of others that also should be looked at. Also I have a habit of losing track of how far I've got 😀.

**Comment by minrk** [comment]
@Carreau wrote a bot to allow anyone to request tag changes via comments, in part for exactly this issue with `needs: info`

**Comment by Carreau** [comment]
The bot can also be configured to only allow some labels to be applied by everyone.
